TEB convolution  in TOAST3

giuspugl opened the pull request:
As suggested by @keskitalo , open a new PR from the developments in  https://github.com/hpc4cmb/toast/pull/505 

This implementation follows the approach implemented by Davide Maino. We perform  3 convolutions,     obtained by splitting both the sky into their {T,E,B} components and the beam alm     into the {T, P}   ones, with P being defined as `P=b_{lm}^E + b_{lm}^B`     The convolution  is performed by  coupling each IQU component of the signal propertly as:
    `skyT_lm * beamT_lm, skyE_lm * Re{beamP_lm}, skyB_lm * Im{beamP_lm}`. In this way, we are able to account for the polarized components of the simulated GRASP beams that were neglected with the Operator in https://github.com/hpc4cmb/toast/blob/38a5b2c5c79500d75811d23807a7d27f557d5818/src/toast/ops/conviqt.py#L571 


For extra details please refer to [this note ](https://giuspugl.github.io/reports/Notes_TEB_convolution.html)

keskitalo commented on src/toast/ops/conviqt.py:
```diff
@@ -658,3 +658,97 @@ def _exec(self, data, detectors=None, **kwargs):
+            convolved_data = self.convolve(sky, beam_T, detector, pnt, det, verbose)
+
+            del pnt,
+            # Q-beam convolution
```
I presume the comment should say `E-beam convolution`?

giuspugl replied:
It is the convolution with the E-beam, resulting in the Q signal

mreineck replied:
Sorry for interrupting with stupid questions: but don't you need both E and B components to compute the Q signal? There is no one-to-one correspondence between E and Q (and B and U), only between E/B and Q/U together, correct?
If you have a pointer where I can read more about this, I'd be really grateful.

giuspugl replied:
Correct the sky E and B alms are convolved with the _polarized _ beam  and both participate to the Q and U  signals. 
have a look at  [this note](https://giuspugl.github.io/reports/Notes_TEB_convolution.html) for some reference

mreineck replied:
Thanks!
So if I understand correctly this still makes some assumptions about the beam, and if we ever wnt to tackle the fully general problem, we have to look at the expressions in https://arxiv.org/abs/2012.10437v1 . I'm still working on that, but I honestly have no idea when that will be ready...
How far will the current approximations take us in terms of LiteBIRD simulations? Will they be good enough for all practical purposes?

giuspugl replied:
The improvements apply mainly to polarization data , now we can include the polarized component of the grasp beam and do not assume it to be the same as the unpolarized one.  As a result this approach yields to lower side lobe pick up than the latter approximation .

keskitalo commented on src/toast/tests/ops_sim_tod_conviqt.py:
```diff
@@ -282,6 +282,89 @@ def test_sim(self):
+
+        # Study the map on the root process
+
+        toast_bin_path = os.path.join(self.outdir, "toast_bin.fits")
```
I would feel more confident if this version of the operator was compared to another version in a case where we expect identical results.

giuspugl replied:
sounds good ! will implment a test

giuspugl replied:
The tests have been reviewed there are major changes involving mainly the scanning strategy (to solve for QU pixels we need more >3 hits), the way the input beam is simulated. 
See  (https://github.com/hpc4cmb/toast/blob/aa03ca92dad37a57a5c5f6afabd65fb61bdc9787/src/toast/tests/ops_sim_tod_conviqt.py#L35)

giuspugl replied:
The operators `SimWeightedConviqt` and `SimTEBConviqt` are tested and compared with `SimConviqt` as they are expect to produce exactly the same convolution outputs in the absence of  spinning HWP

keskitalo commented on src/toast/ops/conviqt.py:
```diff
@@ -658,3 +658,97 @@ def _exec(self, data, detectors=None, **kwargs):
+            del pnt,
+            # Q-beam convolution
+            pnt = self.get_buffer(theta, phi, psi, det, verbose)
+            convolved_data += np.cos(2 * psi_pol) * self.convolve(
```
I am confused about this line here. The note linked in the description says that
``` Q + iU = Sum[(b^E + ib^B)(a^E + ia^B)] * D ```
which implies that
```
Q = Sum[b^E * a^E - b^B * a^B]
U = Sum[b^E * a^B + b^B * a^E]
```
but on this line we seem to be evaluating `Sum[b^E * (a^E + ia^B)]` -- a complex-valued quantity. The issue may be completely semantical but I propose you implement a unit test that compares the standard conviqt operator to this new one. In the absence of a HWP, they should agree.

giuspugl replied:
This is a good point! IT might also linked to https://github.com/hpc4cmb/toast/pull/534#pullrequestreview-869602524 
The way, I implemented this convolution indeed is  as you suggest, meaning that the beams i use in the convolution of Q and U signals are respectively 
```
beam_E = [0, b^E , -b^B ]
beam_B = [0, b^B,b^E]
```

giuspugl replied:
see https://github.com/hpc4cmb/toast/blob/73c758831f3f694ef1d483e53f02de2f928bb929/src/toast/tests/ops_sim_tod_conviqt.py#L113
and https://github.com/hpc4cmb/toast/blob/73c758831f3f694ef1d483e53f02de2f928bb929/src/toast/tests/ops_sim_tod_conviqt.py#L126

keskitalo reviewed (changes requested):
Thanks a lot for wrapping this work up! I got so confused looking at the code that I had to ask for a more extensive unit testing. The current test only shows that the code runs.

mreineck commented:
Do you want me to also port this extension over to the `totalconvolve` operator?

giuspugl commented:
@mreineck this is a good idea! i would firstly merge this PR and then port this to the `totalconvolve` operator !

giuspugl commented:
In #550 
-we corrected a bug in  the implementation of the TEB convolution
- implemented a new test for convolution w/ spinning HWP 

the PR #550 got merged  into this branch and  I think we're ready to final review.

keskitalo commented:
The functional parts of this PR were merged into `toast3` in #552. Thanks a lot for this work, @giuspugl !
